File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.cc
#include<iostream>

using namespace std;

int main() {

	int N;

	int arr[1001] = { 0, };
	int dp[1001] = { 0, };
	int sol = 1;

	cin >> N;

	fill_n(dp, 1001, 1);

	//cout << dp[1000] << endl;

	for (int i = 1; i <= N; i++) {
		cin >> arr[i];
	}

	for (int i = 2; i <= N; i++) {
		for (int j = 1; j < i; j++) {
			if (arr[j] < arr[i]) {
				if (dp[j] + 1 > dp[i])
					dp[i] = dp[j] + 1;
			}
		}
		if (sol < dp[i]) {
			sol = dp[i];
		}
	}

	cout << sol;


	return 0;
}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.*;

class Main{
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        int N = Integer.parseInt(br.readLine());
        int[] dp = new int[N+1];
        String[] str = br.readLine().split(" ");
        int[] arr = new int[N+1];
        int sol=0;

        // 정수화
        for(int i = 1; i <= N; i++){
            arr[i] = Integer.parseInt(str[i-1]);
        }

        dp[1] = 1;
        sol = 1;

        for(int i = 2; i <= N; i++){
            for(int j = i-1; j >= 0; j--){
                if (arr[i] > arr[j]) {
                    if (dp[i] < dp[j] + 1) {
                        dp[i] = dp[j] + 1;
                    }
                }
            }
            if (dp[i] > sol) {
                sol = dp[i];
            }
        }

        bw.write(sol+"");

        bw.flush();
        bw.close();
    }
}
// 1
// 10
// 100 101
// 1000 1001 1010
// 10000 10001 10010 10100 10101